OpenSearch 3.0 PPL Released: Enhanced Log Analytics with Lookup, Join, and Subsearch (#3794)
opensearch-project/project-website: _posts/2025-05-30-Enhanced-Log-Analytics-with-Lookup-Join-and-Subsearch.md

| 2024-04-29T10:00:00Z | jdoe     | login  | success |
| 2024-04-29T10:05:00Z | asmith   | login  | failed  |
| 2024-04-29T10:10:00Z | jdoe     | logout | success |

## Subsearch: Investigate suspicious activity using subsearch

You can use the `subsearch` command to identify users who had failed login attempts and later performed specific actions, such as logging in successfully. This helps surface potentially unauthorized access attempts for further investigation:

```
source=auth_logs 
| where status='failed' AND exists [source=app_logs | where user_id=auth_logs.user_id AND action='login']
```

A sample result appears as follows.

| `timestamp`            | `user_id` | `status` |
| -------------------- | -------- | ------ |
| 2024-04-29T10:05:00Z | asmith   | failed |

## When to use each command

Each command is designed for a specific type of log analysis task. Here's a quick guide to help you choose the right one:

* **`lookup`** – Best for enriching logs with external reference data.
* **`join`** – Ideal for correlating and combining logs from distinct indices.
* **`subsearch`** – Excellent for dynamic query conditions based on results from another query.

> To fully appreciate these new capabilities, it's helpful to understand how they work.

Reviewer: ```suggestion
To fully appreciate these new capabilities, it's helpful to understand how Calcite integration works.
```

> ## How it works: Calcite integration

Reviewer: Reworded

The query processing workflow consists of several transformation stages, converting a query from raw text into various intermediate representations before the final form that is executed. Here's a breakdown of the main stages:

1. **Query parsing**:
   First, the SQL query text is parsed using ANTLR4, which converts the raw query into a parse tree. This tree is then transformed into an AST, which represents the hierarchical syntactic structure of the query in a more abstract form.

2. **Plan conversion**:
   The AST is transformed into Calcite's `RelNode` tree using `RelBuilder` and a series of visitors. `RelNode` is Calcite's internal representation of relational operations. `RelNode` objects form a logical plan tree, where each node represents a relational algebra operation such as filter, project, or join. 

3. **Optimization and transformation**:
   Calcite's optimizer applies various transformation rules to the `RelNode` tree. The optimized tree is then converted into `OpenSearchEnumerableRel` nodes---OpenSearch-specific implementations of Calcite's relational operators, which can be mapped to `Linq4j` expressions.

4. **Code generation and execution**:

>    Linq4j serves as a bridge framework, providing templates and expressions to transform relational operators into executable Java code. This code processes data in a row-by-row manner in memory. For performance optimization, when possible, operators or subtrees of the plan (such as filters and aggregations) are pushed down and converted into native OpenSearch `QueryBuilder` API calls, which are then executed directly by the OpenSearch engine.

Reviewer 2: ```suggestion
   Linq4j serves as a bridge framework, providing templates and expressions for transforming relational operators into executable Java code. This code processes data in a row-by-row manner in memory. For performance optimization, when possible, operators or subtrees of the plan (such as filters and aggregations) are pushed down and converted into native OpenSearch `QueryBuilder` API calls, which are then executed directly by the OpenSearch engine.
```

### How the Calcite optimizer works

> When you run a query, Apache Calcite acts as the query optimizer, transforming your SQL statements into an efficient execution plan. One of the core optimization strategies it applies is **filter pushdown**---a technique that moves filter conditions as close as possible to the data retrieval operations. This reduces the amount of data processed and transferred, leading to significant performance improvements.

Reviewer 2: ```suggestion
When you run a query, Calcite acts as the query optimizer, transforming your SQL statements into an efficient execution plan. One of the core optimization strategies it applies is **filter pushdown**---a technique that moves filter conditions as close as possible to the data retrieval operations. This reduces the amount of data processed and transferred, leading to significant performance improvements.
```

```

In this plan, Calcite determines that the query should filter for `status='failed'` and join the `auth_logs` table with the `user_info` table on the `user_id` field.

#### Physical plan

The physical plan describes how the logical plan operations will be executed. It includes specific steps and strategies, such as sorting, joining methods, and filter pushdown. After optimization, the physical plan might appear as follows:

```
EnumerableCalc(expr#0..5=[{inputs}], proj#0..4=[{exprs}])
  EnumerableMergeJoin(condition=[=($1, $5)], joinType=[left])
    EnumerableSort(sort0=[$1], dir0=[ASC])
      EnumerableCalc(expr#0..9=[{inputs}], proj#0..3=[{exprs}])
        CalciteEnumerableIndexScan(table=[[OpenSearch, auth_logs]], PushDownContext=[[FILTER->=($2, 'failed')], OpenSearchRequestBuilder(sourceBuilder={"from":0,"timeout":"1m","query":{"term":{"status.keyword":{"value":"failed","boost":1.0}}},"sort":[{"_doc":{"order":"asc"}}]}, requestedTotalSize=10000, pageSize=null, startFrom=0)])
    EnumerableSort(sort0=[$1], dir0=[ASC])
      EnumerableCalc(expr#0..7=[{inputs}], proj#0..1=[{exprs}])
        CalciteEnumerableIndexScan(table=[[OpenSearch, user_info]])
```

> In this plan, Calcite pushed down the filter condition `status='failed'` to the `CalciteEnumerableIndexScan` operation for the `auth_logs` table. The filter is then converted to OpenSearch query DSL. This means that filtering occurs directly at the data retrieval stage, minimizing the amount of data loaded into memory and processed by subsequent operations.

Reviewer 2: ```suggestion
In this plan, Calcite pushed down the filter condition `status='failed'` to the `CalciteEnumerableIndexScan` operation for the `auth_logs` table. The filter is then converted to OpenSearch query domain-specific language (DSL). This means that filtering occurs directly at the data retrieval stage, minimizing the amount of data loaded into memory and processed by subsequent operations.
```

> 

Reviewer 2: Line 201, last sentence: "directly at the" => "during the"?

## What's next

The OpenSearch community is continuing to build on the Calcite integration and expand PPL's capabilities. Here's what's planned:

* **Migrate SQL to the Calcite framework** – Unifying SQL and PPL under the Calcite optimizer will streamline query processing and enable more consistent behavior across query languages.

> * **Add more PPL commands** – New commands such as `eventstats` and `streamstats` are planned to support advanced analytical use cases.

Reviewer 2: ```suggestion
* **Add more PPL commands** – New commands, such as `eventstats` and `streamstats`, are planned in order to support advanced analytical use cases.
```

> We invite you to explore the new features in OpenSearch PPL 3.0, share feedback, and contribute to future improvements. Try out the new PPL 3.0 features in your OpenSearch cluster and see how they can improve your log analysis and query workflows. Your input helps shape the direction of the project and ensure it meets the needs of the community.

Reviewer 2: ```suggestion
We invite you to explore the new features in OpenSearch PPL 3.0, share feedback, and contribute to future improvements. Try out the new PPL 3.0 features in your OpenSearch cluster and see how they can improve your log analysis and query workflows. Your input helps shape the direction of the project and ensure that it meets the needs of the community.
```
